File: desktop/ai-app-bridge-cli/test/fixtures/p7-g4a.py
def main(ctx):
    tree = ctx.call("tree", {})
    texts = collect_texts(tree.get("result"))
    ctx.checkpoint("after-tree", {"step": 1})
    tap = ctx.call("tap-text", {"text": "About"})
    after = ctx.call("tree", {})
    verdict = ctx.assert_({
        "name": "license-visible",
        "predicateSummary": "License Notices is on the tree",
        "condition": "License Notices" in collect_texts(after.get("result")),
        "requiredEvidence": [],
        "requireCoverage": "complete",
        "evidence": after.get("evidence"),
    })
    return {
        "passed": verdict.get("verdict") == "passed",
        "tapped": tap.get("execution", {}).get("actionId"),
        "sawAbout": "About" in texts,
    }


def collect_texts(tree):
    found = []
    walk(tree.get("root") if isinstance(tree, dict) else None, found)
    return found


def walk(node, found):
    if not node:
        return
    text = node.get("text")
    if isinstance(text, str):
        found.append(text)
    for child in node.get("children") or []:
        walk(child, found)

File: desktop/ai-app-bridge-cli/test/fixtures/p7-g4a.js
'use strict';

async function main(ctx) {
  const tree = await ctx.call('tree', {});
  const texts = collectTexts(tree.result);
  await ctx.checkpoint('after-tree', { step: 1 });
  const tap = await ctx.call('tap-text', { text: 'About' });
  const after = await ctx.call('tree', {});
  const verdict = await ctx.assert({
    name: 'license-visible',
    predicateSummary: 'License Notices is on the tree',
    condition: collectTexts(after.result).includes('License Notices'),
    requiredEvidence: [],
    requireCoverage: 'complete',
    evidence: after.evidence,
  });
  return {
    passed: verdict.verdict === 'passed',
    tapped: tap.execution.actionId,
    sawAbout: texts.includes('About'),
  };
}

function collectTexts(tree) {
  const found = [];
  walk(tree && tree.root, found);
  return found;
}

function walk(node, found) {
  if (!node) return;
  if (typeof node.text === 'string') found.push(node.text);
  for (const child of node.children || []) walk(child, found);
}

module.exports = { main };
